File: Cocoa/AppFrameworks/Foundation/Safety/NSDictionary+YJSafety.m
//
//  NSDictionary+YJSafety.m
//  YJCocoa
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2019/3/13.
//  Copyright © 2019年 YJCocoa. All rights reserved.
//

#import "NSDictionary+YJSafety.h"
#import "YJNSLog.h"

@implementation NSDictionary (YJSafety)

#define YJNumberGet(value, default) NSNumber *obj = [self objectForKey:aKey];\
if ([obj isKindOfClass:NSNumber.class] || [obj isKindOfClass:NSString.class])\
return obj.value;\
return default;
- (BOOL)boolForKey:(id)aKey {YJNumberGet(boolValue, NO)}
- (NSInteger)integerForKey:(id)aKey {YJNumberGet(integerValue, 0)}
- (CGFloat)floatForKey:(id)aKey {YJNumberGet(floatValue, 0)}

#define YJObjectGet(Cls) Cls *obj = [self objectForKey:aKey];\
if ([obj isKindOfClass:Cls.class]) return obj;\
return [[Cls alloc] init];
- (NSString *)stringForKey:(id)aKey {YJObjectGet(NSString)}
- (NSSet *)setForKey:(id)aKey {YJObjectGet(NSSet)}
- (NSArray *)arrayForKey:(id)aKey {YJObjectGet(NSArray)}
- (NSDictionary *)dictionaryForKey:(id)aKey {YJObjectGet(NSDictionary)}

@end


@implementation NSMutableDictionary (YJSafety)

#define YJObjectGetM(ClsM, Cls) \
ClsM *obj = [self objectForKey:aKey];\
if ([obj isKindOfClass:ClsM.class])  return obj;\
if ([obj isKindOfClass:Cls.class]) obj = obj.mutableCopy;\
else obj = [[ClsM alloc] init];\
[self setObject:obj forKey:aKey];\
return obj;
- (NSMutableSet *)mutableSetForKey:(id)aKey {YJObjectGetM(NSMutableSet, NSSet)}
- (NSMutableArray *)mutableArrayForKey:(id)aKey {YJObjectGetM(NSMutableArray, NSArray)}
- (NSMutableDictionary *)mutableDictionaryForKey:(id)aKey {YJObjectGetM(NSMutableDictionary, NSDictionary)}

#define YJObjectSet \
NSNumber *obj = @(anObject); \
if (obj && aKey) [self setObject:obj forKey:aKey]; \
else YJLogError(@"[Dictionary] set 对应的 key:%@ 或 value:%@ 不存在", obj, aKey);
- (void)setBOOL:(BOOL)anObject forKey:(id)aKey {YJObjectSet}
- (void)setInteger:(NSInteger)anObject forKey:(id)aKey {YJObjectSet}
- (void)setFloat:(CGFloat)anObject forKey:(id)aKey {YJObjectSet}

#define YJObjectSet1(Cls) \
if (anObject && aKey) {\
if ([anObject isKindOfClass:Cls.class]) {[self setObject:anObject forKey:aKey];\
} else {YJLogError(@"[Dictionary] set key:%@ 对应的 value:%@ 类型错误", anObject, aKey);}\
} else {YJLogError(@"[Dictionary] set 对应的 key:%@ 或 value:%@ 不存在", anObject, aKey);}
- (void)setString:(NSString *)anObject forKey:(id)aKey {YJObjectSet1(NSString)}
- (void)setSet:(NSSet *)anObject forKey:(id)aKey {YJObjectSet1(NSSet)}
- (void)setArray:(NSArray *)anObject forKey:(id)aKey {YJObjectSet1(NSArray)}
- (void)setDictionary:(NSDictionary *)anObject forKey:(id)aKey {YJObjectSet1(NSDictionary)}

@end

File: YJCocoa/Classes/AppFrameworks/Foundation/Safety/NSDictionary+YJSafety.swift
//
//  NSDictionary+YJSafety.swift
//  YJCocoa
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2019/5/8.
//  Copyright © 2016-现在 YJCocoa. All rights reserved.
//

import Foundation

public extension NSDictionary {
    
    func bool(forKey key: Any) -> Bool {
        guard let obj = self.object(forKey: key) else {
            return false
        }
        if obj is NSNumber {
            return (obj as! NSNumber).boolValue
        }
        if obj is NSString {
            return (obj as! NSString).boolValue
        }
        return false
    }
    
    func integer(forKey key: Any) -> Int {
        guard let obj = self.object(forKey: key) else {
            return 0
        }
        if obj is NSNumber {
            return (obj as! NSNumber).intValue
        }
        if obj is NSString {
            return (obj as! NSString).integerValue
        }
        return 0
    }
    
    func float(forKey key: Any) -> Float {
        guard let obj = self.object(forKey: key) else {
            return 0.0
        }
        if obj is NSNumber {
            return (obj as! NSNumber).floatValue
        }
        if obj is NSString {
            return (obj as! NSString).floatValue
        }
        return 0.0
    }
    
    func string(forKey key: Any) -> String {
        guard let obj = self.object(forKey: key) else {
            return ""
        }
        if obj is NSNumber {
            return (obj as! NSNumber).stringValue
        }
        if obj is NSString {
            return obj as! String
        }
        return ""
    }
    
    func set(forKey key: Any) -> NSSet {
        guard let obj = self.object(forKey: key) else {
            return NSSet()
        }
        if obj is NSSet {
            return obj as! NSSet
        }
        return NSSet()
    }
    
    func array(forKey key: Any) -> NSArray {
        guard let obj = self.object(forKey: key) else {
            return NSArray()
        }
        if obj is NSArray {
            return obj as! NSArray
        }
        return NSArray()
    }
    
    func dictionary(forKey key: Any) -> NSDictionary {
        guard let obj = self.object(forKey: key) else {
            return NSDictionary()
        }
        if obj is NSDictionary {
            return obj as! NSDictionary
        }
        return NSDictionary()
    }
    
}


public extension NSMutableDictionary {
    
    func mutableSet(forKey key: Any) -> NSMutableSet {
        var obj = self.object(forKey: key)
        if obj is NSMutableSet {
            return obj as! NSMutableSet
        } else if obj is NSSet {
            obj = (obj as! NSSet).mutableCopy()
        } else {
            obj = NSMutableSet()
        }
        self.setObject(obj!, forKey: key as! NSCopying)
        return obj as! NSMutableSet
    }
    
    func mutableArray(forKey key: Any) -> NSMutableArray {
        var obj = self.object(forKey: key)
        if obj is NSMutableArray {
            return obj as! NSMutableArray
        } else if obj is NSArray {
            obj = (obj as! NSArray).mutableCopy()
        } else {
            obj = NSMutableArray()
        }
        self.setArray(obj!, forKey: key as! NSCopying)
        return obj as! NSMutableArray
    }
    
    func mutableDictionary(forKey key: Any) -> NSMutableDictionary {
        var obj = self.object(forKey: key)
        if obj is NSMutableDictionary {
            return obj as! NSMutableDictionary
        } else if obj is NSDictionary {
            obj = (obj as! NSDictionary).mutableCopy()
        } else {
            obj = NSMutableDictionary()
        }
        self.setObject(obj!, forKey: key as! NSCopying)
        return obj as! NSMutableDictionary
    }
    
    func setBool(_ anObject: Bool, forKey aKey: NSCopying) {
        self.setObject(anObject, forKey: aKey)
    }
    
    func setInt(_ anObject: Int, forKey aKey: NSCopying) {
        self.setObject(anObject, forKey: aKey)
    }
    
    func setFloat(_ anObject: Float, forKey aKey: NSCopying) {
        self.setObject(anObject, forKey: aKey)
    }
    
    func setString(_ anObject: Any, forKey aKey: NSCopying) {
        guard let obj = anObject as? NSString else {
            YJLogError("[Dictionary] set key:\(aKey) 对应的 value:\(anObject) 错误")
            return
        }
        self.setObject(obj, forKey: aKey)
    }
    
    func setSet(_ anObject: Any, forKey aKey: NSCopying) {
        guard let obj = anObject as? NSSet else {
            YJLogError("[Dictionary] set key:\(aKey) 对应的 value:\(anObject) 错误")
            return
        }
        self.setObject(obj, forKey: aKey)
    }
    
    func setArray(_ anObject: Any, forKey aKey: NSCopying) {
        guard let obj = anObject as? NSArray else {
            YJLogError("[Dictionary] set key:\(aKey) 对应的 value:\(anObject) 错误")
            return
        }
        self.setObject(obj, forKey: aKey)
    }
    
    func setDictionary(_ anObject: Any, forKey aKey: NSCopying) {
        guard let obj = anObject as? NSDictionary else {
            YJLogError("[Dictionary] set key:\(aKey) 对应的 value:\(anObject) 错误")
            return
        }
        self.setObject(obj, forKey: aKey)
    }
    
}
